File: src/main/java/org/academy/api/client/gui/event/KeyEvent.java
package org.academy.api.client.gui.event;

import net.minecraft.client.input.InputQuirks;

public final class KeyEvent extends InputEvent {
    private final int keyCode;
    private final int scanCode;
    private final int modifiers;

    public KeyEvent(EventType type, int keyCode, int scanCode, int modifiers) {
        super(type);
        if (type != EventType.KEY_PRESSED && type != EventType.KEY_RELEASED) {
            throw new IllegalArgumentException("This constructor is for KEY_PRESSED or KEY_RELEASED events.");
        }
        this.keyCode = keyCode;
        this.scanCode = scanCode;
        this.modifiers = modifiers;
    }

    public int getKeyCode() {
        return keyCode;
    }

    public int getScanCode() {
        return scanCode;
    }

    public int getModifiers() {
        return modifiers;
    }

    public boolean hasAltDown() {
        return (getModifiers() & 4) != 0;
    }

    public boolean hasShiftDown() {
        return (getModifiers() & 1) != 0;
    }

    public boolean hasControlDown() {
        return (getModifiers() & 2) != 0;
    }

    public boolean hasControlDownWithQuirk() {
        return (getModifiers() & InputQuirks.EDIT_SHORTCUT_KEY_MODIFIER) != 0;
    }
}

File: src/main/kotlin/org/academy/api/client/gui/event/KeyEvent.kt
package org.academy.api.client.gui.event

import net.minecraft.client.input.InputQuirks

class KeyEvent(type: EventType, keyCode: Int, scanCode: Int, modifiers: Int) : InputEvent(type) {
    val keyCode: Int
    val scanCode: Int
    val modifiers: Int

    init {
        require(!(type != EventType.KEY_PRESSED && type != EventType.KEY_RELEASED)) { "This constructor is for KEY_PRESSED or KEY_RELEASED events." }
        this.keyCode = keyCode
        this.scanCode = scanCode
        this.modifiers = modifiers
    }

    fun hasAltDown(): Boolean {
        return (this.modifiers and 4) != 0
    }

    fun hasShiftDown(): Boolean {
        return (this.modifiers and 1) != 0
    }

    fun hasControlDown(): Boolean {
        return (this.modifiers and 2) != 0
    }

    fun hasControlDownWithQuirk(): Boolean {
        return (this.modifiers and InputQuirks.EDIT_SHORTCUT_KEY_MODIFIER) != 0
    }
}